### PR TITLE
Implement smartclip on Google AMP AdNetwork

### PR DESCRIPTION
**Pull request by aschubert-sc**
Integrating smartclip on Google amp-ad component / AdNetwork

### CODE REVIEW

**Comment by fkoch-sc** [comment]
Contributor License Agreement (CLA) has been signed.

File: ads/smartclip.js [hunk 1]
```diff
@@ -0,0 +1,35 @@
+/**
+ * Copyright 2015 The AMP HTML Authors. All Rights Reserved.
```
Comment (zhouyx) [inline ads/smartclip.js]: 2016

File: ads/_config.js [hunk 2]
```diff
@@ -316,6 +316,14 @@ export const adConfig = {
     renderStartImplemented: true,
   },
 
+  smartclip: {
+    prefetch: 'https://cdn.smartclip.net/amp/amp.v0.js',
+    preconnect: [
```
Comment (zhouyx) [inline ads/_config.js]: single preconnect url, please do not use array

File: ads/smartclip.js [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {loadScript, validateData} from '../3p/3p';
+
+export function smartclip(global, data) {
```
Comment (zhouyx) [inline ads/smartclip.js]: please add jsdoc, 
https://github.com/ampproject/amphtml/blob/master/ads/google/doubleclick.js#L33
for your example.

File: ads/smartclip.md [hunk 4]
```diff
@@ -0,0 +1,40 @@
+<!---
+Copyright 2015 The AMP HTML Authors. All Rights Reserved.
```
Comment (zhouyx) [inline ads/smartclip.md]: 2016

File: ads/smartclip.md [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+</amp-ad>
+```
+
+## Configuration
+
+For semantics of configuration, please contact [smartclip](mailto:adtech@smartclip.de).
+
+Supported parameters:
+
+All parameters are mandatory, only "extra" is optional.
```
Comment (zhouyx) [inline ads/smartclip.md]: my suggestion would be changing "extra" to code style, use `data-extra`

**Comment by zhouyx** [comment]
@aschubert-sc  Thanks for the Pull Request.
Few comments.  Could you follow our [Developer Guideline](https://github.com/ampproject/amphtml/blob/master/ads/README.md#developer-guidelines-for-a-pull-request) here and add your example ad, and link your doc?

**Comment by fkoch-sc** [comment]
Thanks for the feedback so far @zhouyx. There should be an example ad if you use data-plc="84555". The last adjustment is pushing just now, but everything should be up and running inside 60 minutes from now. I will coordinate with @aschubert-sc tomorrow.

File: examples/ads.amp.html [hunk 6]
```diff
@@ -785,6 +786,13 @@ <h2 id="smartadserver">SmartAdServer ad</h2>
     <div fallback></div>
   </amp-ad>
 
+  <h2 id="smartclip">smartclip</h2>
+  <amp-ad width=400 height=225
+    type="smartclip"
+    data-plc="84555"
+    data-sz="400x320">
```
Comment (zhouyx) [inline examples/ads.amp.html]: please follow the example to add placeholder and fallback.
https://github.com/ampproject/amphtml/blob/master/examples/ads.amp.html#L165

**Comment by zhouyx** [comment]
@fkoch-sc @aschubert-sc I found the ad example you add here is a `no-content` ad. Can you put a `render-start` ad instead?

File: ads/smartclip.js [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+import {loadScript, validateData} from '../3p/3p';
+
+/**
+ * @param {!Window} global
+ * @param {!Object} data
+ */
+export function smartclip(global, data) {
+  /*eslint "google-camelcase/google-camelcase": 0*/
+  global._smartclip_amp = {
+    allowed_data: ['plc', 'sz', 'extra'],
```
Comment (zhouyx) [inline ads/smartclip.js]: you don't have to repeat mandatory_data item in allowed_data array.

**Comment by schubter** [comment]
@zhouyx thanks alot for the feedback, i've updated the code based on that.
the ad should be updated sometime tomorrow morning (berlin time) to deliver a render-start ad

**Comment by aschubert-sc** [comment]
@zhouyx thanks alot for the feedback, i've updated the code based on that.
the ad should be updated sometime tomorrow morning (berlin time) to deliver a render-start ad

**Comment by aschubert-sc** [comment]
the `plc 84555` now delivers a `render-start` ad

File: ads/smartclip.js [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+    data,
+  };
+
+  validateData(data,
+      global._smartclip_amp.mandatory_data, global._smartclip_amp.allowed_data);
+
+  const rand = Math.round(Math.random() * 100000000);
+
+  loadScript(global, 'https://des.smartclip.net/ads?type=dyn&plc='
+    + encodeURI(data.plc) + '&sz=' + encodeURI(data.sz)
```
Comment (zhouyx) [inline ads/smartclip.js]: I see type check failure. Basicly we need used data variable in `ads.extern.js` file. We are having a discussion on whether we should still add these data variable there. But I don't want this to block your pull request. 
So please go ahead and add `data.plc` `data.sz` `data.extra` to `ads.extern.js` file. We will do the refactoring later if we do decide to clean up that file.

**Comment by zhouyx** [comment]
The pull request looks good in general.
@aschubert-sc could you confirm the ad is now working. I still keep seeing ad response timeout or no-content.

**Comment by aschubert-sc** [comment]
yeah @zhouyx i am testing with this ad

File: ads/smartclip.js (hunk 8 above)
Comment (aschubert-sc) [inline ads/smartclip.js]: i could change the jsdoc header to 
` * @param {!{plc: String, sz: String}} data` but not sure if that helps

Comment (zhouyx) [inline ads/smartclip.js]: adding to ads.extern.js should be good enough

**Comment by zhouyx** [comment]
@aschubert-sc I am able to see render-start ad now.  
LGTM from my side. Ping @lannka to take a final look.

**Comment by zhouyx** [comment]
ping @lannka

File: 3p/integration.js [hunk 9]
```diff
@@ -117,6 +118,7 @@ const AMP_EMBED_ALLOWED = {
   plista: true,
   zergnet: true,
   _ping_: true,
+  smartclip: true,
```
Comment (lannka) [inline 3p/integration.js]: please sort it in alphabetic order

File: ads/smartclip.js (hunk 8 above)
Comment (lannka) [inline ads/smartclip.js]: should use `encodeURIComponent`?

File: ads/smartclip.js [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+    + (data.extra ? '&' + data.extra : '') + '&rnd=' + rand);
```
Comment (lannka) [inline ads/smartclip.js]: encode `data.extra` as well?

**Comment by zhouyx** [comment]
@aschubert-sc Please resolve conflicts and good to be merged.

**Comment by zhouyx** [comment]
@aschubert-sc Thanks for the PR. Merged

**Comment by fkoch-sc** [comment]
Good news. Thanks for your feedback, @zhouyx and @lannka.

**Pull request merged**